File: mr/src/main/java/org/apache/mahout/classifier/sequencelearning/hmm/HmmModel.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.mahout.classifier.sequencelearning.hmm;

import java.util.Map;
import java.util.Random;

import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import org.apache.mahout.common.RandomUtils;
import org.apache.mahout.math.DenseMatrix;
import org.apache.mahout.math.DenseVector;
import org.apache.mahout.math.Matrix;
import org.apache.mahout.math.Vector;

/**
 * Main class defining a Hidden Markov Model
 */
public class HmmModel implements Cloneable {

  /** Bi-directional Map for storing the observed state names */
  private BiMap<String,Integer> outputStateNames;

  /** Bi-Directional Map for storing the hidden state names */
  private BiMap<String,Integer> hiddenStateNames;

  /* Number of hidden states */
  private int nrOfHiddenStates;

  /** Number of output states */
  private int nrOfOutputStates;

  /**
   * Transition matrix containing the transition probabilities between hidden
   * states. TransitionMatrix(i,j) is the probability that we change from hidden
   * state i to hidden state j In general: P(h(t+1)=h_j | h(t) = h_i) =
   * transitionMatrix(i,j) Since we have to make sure that each hidden state can
   * be "left", the following normalization condition has to hold:
   * sum(transitionMatrix(i,j),j=1..hiddenStates) = 1
   */
  private Matrix transitionMatrix;

  /**
   * Output matrix containing the probabilities that we observe a given output
   * state given a hidden state. outputMatrix(i,j) is the probability that we
   * observe output state j if we are in hidden state i Formally: P(o(t)=o_j |
   * h(t)=h_i) = outputMatrix(i,j) Since we always have an observation for each
   * hidden state, the following normalization condition has to hold:
   * sum(outputMatrix(i,j),j=1..outputStates) = 1
   */
  private Matrix emissionMatrix;

  /**
   * Vector containing the initial hidden state probabilities. That is
   * P(h(0)=h_i) = initialProbabilities(i). Since we are dealing with
   * probabilities the following normalization condition has to hold:
   * sum(initialProbabilities(i),i=1..hiddenStates) = 1
   */
  private Vector initialProbabilities;


  /**
   * Get a copy of this model
   */
  @Override
  public HmmModel clone() {
    HmmModel model = new HmmModel(transitionMatrix.clone(), emissionMatrix.clone(), initialProbabilities.clone());
    if (hiddenStateNames != null) {
      model.hiddenStateNames = HashBiMap.create(hiddenStateNames);
    }
    if (outputStateNames != null) {
      model.outputStateNames = HashBiMap.create(outputStateNames);
    }
    return model;
  }

  /**
   * Assign the content of another HMM model to this one
   *
   * @param model The HmmModel that will be assigned to this one
   */
  public void assign(HmmModel model) {
    this.nrOfHiddenStates = model.nrOfHiddenStates;
    this.nrOfOutputStates = model.nrOfOutputStates;
    this.hiddenStateNames = model.hiddenStateNames;
    this.outputStateNames = model.outputStateNames;
    // for now clone the matrix/vectors
    this.initialProbabilities = model.initialProbabilities.clone();
    this.emissionMatrix = model.emissionMatrix.clone();
    this.transitionMatrix = model.transitionMatrix.clone();
  }

  /**
   * Construct a valid random Hidden-Markov parameter set with the given number
   * of hidden and output states using a given seed.
   *
   * @param nrOfHiddenStates Number of hidden states
   * @param nrOfOutputStates Number of output states
   * @param seed             Seed for the random initialization, if set to 0 the current time
   *                         is used
   */
  public HmmModel(int nrOfHiddenStates, int nrOfOutputStates, long seed) {
    this.nrOfHiddenStates = nrOfHiddenStates;
    this.nrOfOutputStates = nrOfOutputStates;
    this.transitionMatrix = new DenseMatrix(nrOfHiddenStates, nrOfHiddenStates);
    this.emissionMatrix = new DenseMatrix(nrOfHiddenStates, nrOfOutputStates);
    this.initialProbabilities = new DenseVector(nrOfHiddenStates);
    // initialize a random, valid parameter set
    initRandomParameters(seed);
  }

  /**
   * Construct a valid random Hidden-Markov parameter set with the given number
   * of hidden and output states.
   *
   * @param nrOfHiddenStates Number of hidden states
   * @param nrOfOutputStates Number of output states
   */
  public HmmModel(int nrOfHiddenStates, int nrOfOutputStates) {
    this(nrOfHiddenStates, nrOfOutputStates, 0);
  }

  /**
   * Generates a Hidden Markov model using the specified parameters
   *
   * @param transitionMatrix     transition probabilities.
   * @param emissionMatrix       emission probabilities.
   * @param initialProbabilities initial start probabilities.
   * @throws IllegalArgumentException If the given parameter set is invalid
   */
  public HmmModel(Matrix transitionMatrix, Matrix emissionMatrix, Vector initialProbabilities) {
    this.nrOfHiddenStates = initialProbabilities.size();
    this.nrOfOutputStates = emissionMatrix.numCols();
    this.transitionMatrix = transitionMatrix;
    this.emissionMatrix = emissionMatrix;
    this.initialProbabilities = initialProbabilities;
  }

  /**
   * Initialize a valid random set of HMM parameters
   *
   * @param seed seed to use for Random initialization. Use 0 to use Java-built-in-version.
   */
  private void initRandomParameters(long seed) {
    Random rand;
    // initialize the random number generator
    if (seed == 0) {
      rand = RandomUtils.getRandom();
    } else {
      rand = RandomUtils.getRandom(seed);
    }
    // initialize the initial Probabilities
    double sum = 0; // used for normalization
    for (int i = 0; i < nrOfHiddenStates; i++) {
      double nextRand = rand.nextDouble();
      initialProbabilities.set(i, nextRand);
      sum += nextRand;
    }
    // "normalize" the vector to generate probabilities
    initialProbabilities = initialProbabilities.divide(sum);

    // initialize the transition matrix
    double[] values = new double[nrOfHiddenStates];
    for (int i = 0; i < nrOfHiddenStates; i++) {
      sum = 0;
      for (int j = 0; j < nrOfHiddenStates; j++) {
        values[j] = rand.nextDouble();
        sum += values[j];
      }
      // normalize the random values to obtain probabilities
      for (int j = 0; j < nrOfHiddenStates; j++) {
        values[j] /= sum;
      }
      // set this row of the transition matrix
      transitionMatrix.set(i, values);
    }

    // initialize the output matrix
    values = new double[nrOfOutputStates];
    for (int i = 0; i < nrOfHiddenStates; i++) {
      sum = 0;
      for (int j = 0; j < nrOfOutputStates; j++) {
        values[j] = rand.nextDouble();
        sum += values[j];
      }
      // normalize the random values to obtain probabilities
      for (int j = 0; j < nrOfOutputStates; j++) {
        values[j] /= sum;
      }
      // set this row of the output matrix
      emissionMatrix.set(i, values);
    }
  }

  /**
   * Getter Method for the number of hidden states
   *
   * @return Number of hidden states
   */
  public int getNrOfHiddenStates() {
    return nrOfHiddenStates;
  }

  /**
   * Getter Method for the number of output states
   *
   * @return Number of output states
   */
  public int getNrOfOutputStates() {
    return nrOfOutputStates;
  }

  /**
   * Getter function to get the hidden state transition matrix
   *
   * @return returns the model's transition matrix.
   */
  public Matrix getTransitionMatrix() {
    return transitionMatrix;
  }

  /**
   * Getter function to get the output state probability matrix
   *
   * @return returns the models emission matrix.
   */
  public Matrix getEmissionMatrix() {
    return emissionMatrix;
  }

  /**
   * Getter function to return the vector of initial hidden state probabilities
   *
   * @return returns the model's init probabilities.
   */
  public Vector getInitialProbabilities() {
    return initialProbabilities;
  }

  /**
   * Getter method for the hidden state Names map
   *
   * @return hidden state names.
   */
  public Map<String, Integer> getHiddenStateNames() {
    return hiddenStateNames;
  }

  /**
   * Register an array of hidden state Names. We assume that the state name at
   * position i has the ID i
   *
   * @param stateNames names of hidden states.
   */
  public void registerHiddenStateNames(String[] stateNames) {
    if (stateNames != null) {
      hiddenStateNames = HashBiMap.create();
      for (int i = 0; i < stateNames.length; ++i) {
        hiddenStateNames.put(stateNames[i], i);
      }
    }
  }

  /**
   * Register a map of hidden state Names/state IDs
   *
   * @param stateNames <String,Integer> Map that assigns each state name an integer ID
   */
  public void registerHiddenStateNames(Map<String, Integer> stateNames) {
    if (stateNames != null) {
      hiddenStateNames = HashBiMap.create(stateNames);
    }
  }

  /**
   * Lookup the name for the given hidden state ID
   *
   * @param id Integer id of the hidden state
   * @return String containing the name for the given ID, null if this ID is not
   *         known or no hidden state names were specified
   */
  public String getHiddenStateName(int id) {
    if (hiddenStateNames == null) {
      return null;
    }
    return hiddenStateNames.inverse().get(id);
  }

  /**
   * Lookup the ID for the given hidden state name
   *
   * @param name Name of the hidden state
   * @return int containing the ID for the given name, -1 if this name is not
   *         known or no hidden state names were specified
   */
  public int getHiddenStateID(String name) {
    if (hiddenStateNames == null) {
      return -1;
    }
    Integer tmp = hiddenStateNames.get(name);
    return tmp == null ? -1 : tmp;
  }

  /**
   * Getter method for the output state Names map
   *
   * @return names of output states.
   */
  public Map<String, Integer> getOutputStateNames() {
    return outputStateNames;
  }

  /**
   * Register an array of hidden state Names. We assume that the state name at
   * position i has the ID i
   *
   * @param stateNames state names to register.
   */
  public void registerOutputStateNames(String[] stateNames) {
    if (stateNames != null) {
      outputStateNames = HashBiMap.create();
      for (int i = 0; i < stateNames.length; ++i) {
        outputStateNames.put(stateNames[i], i);
      }
    }
  }

  /**
   * Register a map of hidden state Names/state IDs
   *
   * @param stateNames <String,Integer> Map that assigns each state name an integer ID
   */
  public void registerOutputStateNames(Map<String, Integer> stateNames) {
    if (stateNames != null) {
      outputStateNames = HashBiMap.create(stateNames);
    }
  }

  /**
   * Lookup the name for the given output state id
   *
   * @param id Integer id of the output state
   * @return String containing the name for the given id, null if this id is not
   *         known or no output state names were specified
   */
  public String getOutputStateName(int id) {
    if (outputStateNames == null) {
      return null;
    }
    return outputStateNames.inverse().get(id);
  }

  /**
   * Lookup the ID for the given output state name
   *
   * @param name Name of the output state
   * @return int containing the ID for the given name, -1 if this name is not
   *         known or no output state names were specified
   */
  public int getOutputStateID(String name) {
    if (outputStateNames == null) {
      return -1;
    }
    Integer tmp = outputStateNames.get(name);
    return tmp == null ? -1 : tmp;
  }

}

File: math-scala/src/main/scala/org/apache/mahout/classifier/sequencelearning/hmm/HMMModel.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.mahout.classifier.sequencelearning.hmm

import org.apache.mahout.math._

import org.apache.mahout.math.{drm, scalabindings}

import scalabindings._
import scalabindings.RLikeOps._
import drm._
import scala.language.asInstanceOf
import scala.collection._
import JavaConversions._
import scala.util.Random
import org.apache.mahout.common.RandomUtils

/**
 *
 * @param numHiddenStates number of hidden states
 * @param numOutputStates number of output states
 */
class HMMModel(val numberOfHiddenStates: Int,
               val numberOfOutputSymbols: Int,
	       var transitionMatrix: Matrix = null,
	       var emissionMatrix: Matrix = null,
  	       var initialProbabilities: Vector = null)  extends java.io.Serializable {

  var cumulativesInitialized = false
  var cumulativeTransitionMatrix:Matrix = _
  var cumulativeEmissionMatrix:Matrix = _
  var cumulativeInitialProbabilities:Vector = _
  var validateModel:Boolean = false

  if (initialProbabilities == null) {
    initialProbabilities = new DenseVector(numberOfHiddenStates)
  } else {
    validateModel = true
  }

  if (transitionMatrix == null ) {
    transitionMatrix = new DenseMatrix(numberOfHiddenStates, numberOfHiddenStates)
  }

  if (emissionMatrix == null) {
    emissionMatrix = new DenseMatrix(numberOfHiddenStates, numberOfOutputSymbols)
  }

  if (validateModel) {
    validate()
  }

  def getNumberOfHiddenStates: Int = {
    numberOfHiddenStates
  }

  def getNumberOfObservableSymbols: Int = {
    numberOfOutputSymbols
  }

  def getInitialProbabilities: Vector = {
    initialProbabilities
  }

  def getEmissionMatrix: Matrix = {
    emissionMatrix
  }

  def getTransitionMatrix: Matrix = {
    transitionMatrix
  }

  def getCumulativeTransitionMatrix: Matrix = {
    if (cumulativesInitialized == false) {
      cumulativeTransitionMatrix = new DenseMatrix(numberOfHiddenStates, numberOfHiddenStates)
      for (i <- 0 until numberOfHiddenStates) {
        var sum:Double = 0;
        for (j <- 0 until numberOfHiddenStates) {
          sum = sum + transitionMatrix.getQuick(i, j)
          cumulativeTransitionMatrix.setQuick(i, j, sum)
        }
        cumulativeTransitionMatrix.setQuick(i, numberOfHiddenStates - 1, 1.0)
      }
    }

    cumulativeTransitionMatrix
  }

  def getCumulativeEmissionMatrix: Matrix = {
    if (cumulativesInitialized == false) {
      cumulativeEmissionMatrix = new DenseMatrix(numberOfHiddenStates, numberOfOutputSymbols)
      for (i <- 0 until numberOfHiddenStates) {
        var sum:Double = 0;
        for (j <- 0 until numberOfOutputSymbols) {
          sum = sum + emissionMatrix.getQuick(i, j)
          cumulativeEmissionMatrix.setQuick(i, j, sum)
        }
        cumulativeEmissionMatrix.setQuick(i, numberOfOutputSymbols - 1, 1.0)
      }
    }

    cumulativeEmissionMatrix
  }

  def getCumulativeInitialProbabilities: Vector = {
    if (cumulativesInitialized == false) {
      cumulativeInitialProbabilities = new DenseVector(numberOfHiddenStates)
      var sum:Double = 0;
      for (i <- 0 until numberOfHiddenStates) {
        sum = sum + initialProbabilities.getQuick(i)
        cumulativeInitialProbabilities.setQuick(i, sum)
      }
      cumulativeInitialProbabilities.setQuick(numberOfHiddenStates - 1, 1.0)
    }

    cumulativeInitialProbabilities
  }

  def printModel(): Unit = {
    println("Transition Matrix:")
    println(getTransitionMatrix)
    println("Emission Matrix:")
    println(getEmissionMatrix)
    println("Initial Probabilities Vector:")
    println(getInitialProbabilities)
  }

  def initModelWithRandomParameters(seed:Long): Unit = {
    var rand:Random = RandomUtils.getRandom()
    if (seed != 0) {
      rand = RandomUtils.getRandom(seed)
    }

    // initialize the initial Probabilities
    var sum:Double = 0 
    for (i <- 0 to numberOfHiddenStates - 1) {
      val nextRand:Double = rand.nextDouble();
      initialProbabilities.setQuick(i, nextRand);
      sum += nextRand;
    }
    // "normalize" the vector to generate probabilities
    for (i <- 0 to numberOfHiddenStates - 1) {
      initialProbabilities.setQuick(i, initialProbabilities.getQuick(i)/sum);
    }

    // initialize the transition matrix
    for (i <- 0 to numberOfHiddenStates -1) {
      sum = 0
      for (j <- 0 to numberOfHiddenStates - 1) {
        transitionMatrix.setQuick(i, j, rand.nextDouble())
        sum += transitionMatrix.getQuick(i, j)
      }
      // normalize the random values to obtain probabilities
      for (j <- 0 to numberOfHiddenStates - 1) {
        transitionMatrix.setQuick(i, j, transitionMatrix.getQuick(i, j)/sum)
      }
    }

    // initialize the output matrix
    for (i <- 0 to numberOfHiddenStates - 1) {
      sum = 0
      for (j <- 0 to numberOfOutputSymbols - 1) {
        emissionMatrix.setQuick(i, j, rand.nextDouble())
        sum += emissionMatrix.getQuick(i, j)
      }
      // normalize the random values to obtain probabilities
      for (j <- 0 to numberOfOutputSymbols - 1) {
        emissionMatrix.setQuick(i, j, emissionMatrix.getQuick(i, j)/sum)
      }
    }
  }

  /**
   * Write a trained model to the filesystem as a series of DRMs
   * @param pathToModel Directory to which the model will be written
   */
  def dfsWrite(pathToModel: String)(implicit ctx: DistributedContext): Unit = {
    val fullPathToModel = pathToModel + HMMModel.modelBaseDirectory
    drmParallelize(sparse(svec((0, numberOfHiddenStates)::Nil))).dfsWrite(fullPathToModel + "/numberOfHiddenStatesDrm.drm")
    drmParallelize(sparse(svec((0, numberOfOutputSymbols)::Nil))).dfsWrite(fullPathToModel + "/numberOfOutputSymbolsDrm.drm")
    drmParallelize(transitionMatrix).dfsWrite(fullPathToModel + "/transitionMatrixDrm.drm")
    drmParallelize(emissionMatrix).dfsWrite(fullPathToModel + "/emissionMatrixDrm.drm")
    drmParallelize(sparse(initialProbabilities)).dfsWrite(fullPathToModel + "/initialProbabilitiesDrm.drm")
  }

  /** Model Validation */
  def validate() {
    assert(numberOfHiddenStates > 0, "number of hidden states has to be greater than 0.")
    assert(numberOfOutputSymbols > 0, "number of output symbols has to be greater than 0.")
    assert(transitionMatrix.rowSize() == numberOfHiddenStates, "number of rows of transition matrix should be equal to number of hidden states")
    assert(transitionMatrix.columnSize() == numberOfHiddenStates, "number of columns of transition matrix should be equal to number of hidden states")
    assert(emissionMatrix.rowSize() == numberOfHiddenStates, "number of rows of emission matrix should be equal to number of hidden states")
    assert(emissionMatrix.columnSize() == numberOfOutputSymbols, "number of columns of emission matrix should be equal to number of output symbols")
    assert(initialProbabilities.size() == numberOfHiddenStates, "number of entries of initial probabilities vector should be equal to number of hidden states")

    var sum:Double = 0
    for (i <- 0 until initialProbabilities.size()) {
      assert(initialProbabilities.getQuick(i) >= 0, "Initial probability of state is negative")
      sum += initialProbabilities.getQuick(i)
    }

    assert(Math.abs(sum - 1) <= 0.000001, "Initial probabilities do not add up to 1")

    for (i <- 0 until numberOfHiddenStates) {
      sum = 0
      for (j <- 0 until numberOfHiddenStates) {
        assert(transitionMatrix.getQuick(i, j) >= 0, "transition probability is negative")
        sum += transitionMatrix.getQuick(i, j)
      }
      assert(Math.abs(sum - 1) <= 0.000001, "transition probabilities do not add up to 1")
    }

    for (i <- 0 until numberOfHiddenStates) {
      sum = 0
      for (j <- 0 until numberOfOutputSymbols) {
        assert(emissionMatrix.getQuick(i, j) >= 0, "emission probability is negative")
        sum += emissionMatrix.getQuick(i, j)
      }
      assert(Math.abs(sum - 1) <= 0.000001, "emission probabilities do not add up to 1")
    }
  }
}

object HMMModel extends java.io.Serializable {

  val modelBaseDirectory = "/hiddenMarkovModel"
  
  def dfsRead(pathToModel: String)(implicit ctx: DistributedContext): HMMModel = {
    // read from a base directory for all drms
    val fullPathToModel = pathToModel + HMMModel.modelBaseDirectory
    val numberOfHiddenStatesDrm = drmDfsRead(fullPathToModel + "/numberOfHiddenStatesDrm.drm").checkpoint(CacheHint.MEMORY_ONLY)
    val numberOfHiddenStates = numberOfHiddenStatesDrm.collect(0, 0).toInt
    numberOfHiddenStatesDrm.uncache()
    val numberOfOutputSymbolsDrm = drmDfsRead(fullPathToModel + "/numberOfOutputSymbolsDrm.drm").checkpoint(CacheHint.MEMORY_ONLY)
    val numberOfOutputSymbols = numberOfOutputSymbolsDrm.collect(0, 0).toInt
    numberOfOutputSymbolsDrm.uncache()
    val transitionMatrixDrm = drmDfsRead(fullPathToModel + "/transitionMatrixDrm.drm").checkpoint(CacheHint.MEMORY_ONLY)
    val transitionMatrix = transitionMatrixDrm.collect
    transitionMatrixDrm.uncache()
    val emissionMatrixDrm = drmDfsRead(fullPathToModel + "/emissionMatrixDrm.drm").checkpoint(CacheHint.MEMORY_ONLY)
    val emissionMatrix = emissionMatrixDrm.collect
    emissionMatrixDrm.uncache()
    val initialProbabilitiesDrm = drmDfsRead(fullPathToModel + "/initialProbabilitiesDrm.drm").checkpoint(CacheHint.MEMORY_ONLY)
    val initialProbabilities = initialProbabilitiesDrm.collect(0, ::)
    initialProbabilitiesDrm.uncache()
    val model: HMMModel = new HMMModel(numberOfHiddenStates, numberOfOutputSymbols, transitionMatrix, emissionMatrix, initialProbabilities)

    model
  }
}
